react/components/NumericStepper/index.js

```diff
@@ -0,0 +1,223 @@
+import PropTypes from 'prop-types'
+import React from 'react'
+
+const normalizeMin = min => min == null ? -Infinity : min
+const normalizeMax = max => max == null ? Infinity : max
+
+const validateValue = (value, min, max, defaultValue) => {
+  min = normalizeMin(min)
+  max = normalizeMax(max)
+
+  if (isNaN(value) || value == null) {
+    if (defaultValue < min) return min
+    if (defaultValue > max) return max
+    return defaultValue
+  } else if (value < min) {
+    return min
+  } else if (value > max) {
+    return max
+  }
+  return parseInt(value, 10)
+}
+
+const validateDisplayValue = (value, min, max) => {
+  min = normalizeMin(min)
+  max = normalizeMax(max)
+
+  const parsedValue = parseInt(value, 10)
+
+  if (value === '') {
+    return value
+  }
+  if (value === '-' && min < 0) {
+    return value
+  }
+  if (isNaN(parsedValue)) {
+    return ''
+  }
+  if (parsedValue < min) {
+    return min
+  }
+  if (parsedValue > max) {
+    return max
+  }
+  return parsedValue
+}
+
+export default class NumericStepper extends React.Component {
+  state = {
+    inputFocused: false,
+    // used for comparison whether to trigger onChange or not
+    value: 0,
+    // used for temporarily invalid values during typing--specifically, when it's empty
+    displayValue: 0,
+  }
+
+  static getDerivedStateFromProps(props, state) {
+    const { value, minValue, maxValue, defaultValue } = props
+
+    const validatedValue = validateValue(value, minValue, maxValue, defaultValue)
+
+    return {
+      value: validatedValue,
+      ...(!state.inputFocused && {
+        displayValue: validateDisplayValue(value, minValue, maxValue),
+      }),
+    }
+  }
+
+  changeValue = value => {
+    const parsedValue = parseInt(value, 10)
+
+    const { minValue, maxValue, defaultValue, onChange } = this.props
+
+    const validatedValue = validateValue(parsedValue, minValue, maxValue, defaultValue)
+
+    const displayValue = validateDisplayValue(value, minValue, maxValue)
+
+    this.setState({
+      value: validatedValue,
+      displayValue,
+    })
+
+    if (this.state.value !== validatedValue) {
+      onChange && onChange(validatedValue)
```

Comment:
Acho estranho o onChange desse componente receber value ao inves de um evento como todo os outros.

Ver:
> Keep prop APIs parallel
http://jxnblk.com/writing/posts/defining-component-apis-in-react/

Reply:
Bem notado

Reply:
@klzns Apesar que esse é um evento criado pelo componente, não um evento nativo. Crio um evento manualmente?

Reply:
Nao da pra repassar o evento original e mutar o value?

Reply:
@klzns Eu posso, mas pode ser tanto evento de change do input quanto de click nos botões. Tanto o tipo do evento, quanto os targets, mudam.

Vou fazer isso, mas talvez nao seja o ideal.